File: go/internal/monorepo/filter_ignored.go
package monorepo

import (
	"fmt"
	"path/filepath"

	"github.com/bmatcuk/doublestar/v4"
)

// FilterIgnoredFiles filters out files that match ignore patterns.
// Uses matchBase behavior: patterns without path separators match against the basename.
func FilterIgnoredFiles(files []string, patterns []string, verbose bool) []string {
	var result []string

	for _, file := range files {
		ignored := false
		var matchedPattern string

		for _, pattern := range patterns {
			matched := matchFile(file, pattern)
			if matched {
				ignored = true
				matchedPattern = pattern
				break
			}
		}

		if ignored {
			if verbose {
				fmt.Printf("  - ~~%s~~ (ignored by pattern %q)\n", file, matchedPattern)
			}
		} else {
			result = append(result, file)
		}
	}

	return result
}

// matchFile checks if a file matches a pattern, using matchBase for patterns without slashes.
func matchFile(file, pattern string) bool {
	// If pattern has no path separator, match against basename (matchBase behavior)
	hasSlash := false
	for _, c := range pattern {
		if c == '/' || c == '\\' {
			hasSlash = true
			break
		}
	}

	if !hasSlash {
		base := filepath.Base(file)
		if matched, _ := doublestar.PathMatch(pattern, base); matched {
			return true
		}
	}

	// Try full path match
	if matched, _ := doublestar.PathMatch(pattern, file); matched {
		return true
	}

	// Try with **/ prefix for patterns with path separators
	if hasSlash {
		if matched, _ := doublestar.PathMatch("**/"+pattern, file); matched {
			return true
		}
	}

	return false
}

File: rust/src/monorepo/filter_ignored.rs
use super::path_included::match_with_base;
use crate::log_verbose;

/// Filter out file paths that match any of the ignore patterns.
/// Uses matchBase: true behavior (patterns without '/' match against basename).
pub fn filter_ignored_files(file_paths: &[String], ignore_patterns: &[String]) -> Vec<String> {
    file_paths
        .iter()
        .filter(|path| {
            for pattern in ignore_patterns {
                if match_with_base(path, pattern) {
                    log_verbose!("  - ~~{path}~~ (ignored by pattern \"{pattern}\")");
                    return false;
                }
            }
            true
        })
        .cloned()
        .collect()
}
